### PR TITLE
COMP3710 StyleGAN Implementation for ADNI Brain Image Generation

### PR DESCRIPTION
**Pull request by MRMATTHEWLGIT**
Hello COMP3710 Team,

This fork of the `topic-recognition` branch contains an implementation of the StyelGAN deep neural network architecture. Which was implemented to try and solve the problem of both generating brain scan images based on the ADNI dataset, while also having the Discriminator and Generator be able to distinguish between the two classes, being Cognitive Normal (CN) and Alzheimer's Disease (AD).  This also includes the functionality to plot the UMAP Embeddings based on the feature map output of the Generator, excited by the class labels it receives. 

For further information regarding the entire implementation of this StyleGAN architecture, it's performance, and it's dependencies, please don't hesitate to view the `README.md` file contained within this pull request. For all files pertaining to the StyleGAN model, including the `README.md`, go to the following directory `recognition/StyleGAN-46988133` within the `topic-recognition` branch.

Here is a brief summary of each item contained within this pull request:

* `Assets\`: Used to store all figures and images required for the `README.md` file. This includes training loss plots, real vs fake image comparisons and UMAP Embedding plots.
* `dataset.py`: Loads the required ADNI dataset images, and performs simple torchvision transforms on them to get them to the desired resolution and normalisation range. Follow the dataset structure comments within this file to ensure your ADNI dataset is structured in the required way.
* `environment.yml`: Contains the definitions for the Conda environment used to run this program. This can be used to install the same Conda environment used during development of this StyleGAN model. Information for how to perform this action is explained in the `README.md`, in the section labelled 'How To Run'.
* `hyperparameters.py`: Contains all hyperparameters and configuration constants used to train and visualise the StyleGAN model. To change the dataset location to one that you have, change the `ROOT` constant within this file. 
* `modules.py`: Contains the entire implementation of this StyleGAN, including the classes for the Discriminator, Generator and Mapping Network. Plus any helper functions and classes necessary to them.
* `README.md`: Contains helpful documentation about the entire StyleGAN development process, results, dependencies and even possible future improvements that could be applied. If you have any questions after reading this pull request description, please head there for further information.
* `predict.py`: Loads both the Discriminator and Generator models after being trained through various resolutions, and is used to validate that they work. This is done by plotting a UMAP Embedding based on the Generator's feature map output, and also a comparison of the real ADNI dataset images compared to the ones created by the Generator.
* `train.py`: Implements the main training loop for this StyleGAN implementation, which uses a baseline ProGAN model to incrementally increase the resolution of the training image. Once each resolution is trained fully, various plots and figures are saved to the `/saved_output` folder.
* `utils.py`: Contains a helper function , specifically one to help plot the necessary figures after every resolution is trained within the training loop.

Thank you, 

Matthew Lockett [46988133]

### CODE REVIEW

**Comment by Uto0op** [comment]
Recognition problem: StyleGAN on ADNI ---- Hard Difficulty

1. Algorithm solves the problem: (4.5 Marks)
Somewhat not reasonably clear generated images (32x32) -0.5;

2. Algorithm demonstrated and functions: (3 Marks)

4. Design: (1 Mark)
Modularized design, good programming practice using utils script

5. Commenting: (1 Mark)
Sufficient docstrings and detailed comments

6. Algorithm above Normal Difficulty: (5 Marks)

7. Algorithm is Hard difficulty: (5 Marks)

**Suggestions**:
- Firstly, I do think the progressive training framework adapted from ProGAN should work, yet you mentioned:

> In terms of image generation this StyleGAN implementation was not able to fully progress through the chain of ProGAN image sizes, being 8x8, 32x32, 64x64, 128x128 and lastly 256x256.

It would be good if you can clarify the reason. If this is due to memory issue, you can try to decrease the batch size. Also, the final 256x256 might not be a good idea as the original image size is 256x240. Anyway, I'm afraid 32x32 is not sufficient to present "reasonably clear images", much information would be lost and performance would be jeopardized (think about it, images of higher resolutions were not essentially used to train the model given this progressive training approach). I'm happy to give the 0.5 mark back if you can justify;
- It seems you only used the training set to train the model, while you could make use of both training and test sets since there's no testing for this task ---- this might improve the current performance;
- The UMAP was plotted on the feature map extracted from the last layer of the generator, while ideally it should be plotted on the style space (see the reference paper from the task sheet). I'm not taking marks off for it as it seems you are aware of what you are doing based on your justification. I just need to point out that this is not the intention of this task, and by plotting the UMAP on the style code, it's expected to observe better clustering as your model is conditional;
- Scripts `environment.yml` and `hyperparameters.py` are not needed for this project, but good practice for the purpose of modularization and reproducibility;
- For the Future Improvements, I'm still convinced progressive training should be beneficial given sufficient resources, and the Wasserstein loss makes sense if it can converge for higher resolutions;
- Very good attempt to introduce conditions into the StyleGAN;
- Very elaborative documentation and good referencing practice.
- Very good project, showing effort put :)

**Comment by yexincheng** [comment]
**This is an initial inspection, no action is required at this point**
1. **Solves problem**: Nice documentation and introduction of the network. All generated images are the same, which is unreasonable. The loss plot of 32*32 resolution is provided; it would be better to have the loss plot for each resolution. Nice UMAP plot, clear two clusters. Nice records for experiment steps.
2. **Implementation**: Very well structured. 
3. **Design**: Good
4. **Comment**: Very nice and clear, sufficient and detailed comments.
5. **Difficulty**: Hard

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Low image quality.</td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. StyleGAN (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. Some images inconsistancies.</td><td>-1</td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). ON time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-2</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Good work!
